File: src/services/clustering.ts
/**
 * K-means Clustering Service
 * Groups students by geographic proximity for vehicle assignment
 * Based on MATLAB VRP code logic
 */

import type { Coordinates } from "@/lib/coordinates";
import { haversineDistance } from "@/lib/coordinates";

export interface ClusterablePoint {
    id: string;
    coordinates: Coordinates;
    disabilityType: "Sw" | "So";
}

export interface Cluster {
    centroid: Coordinates;
    points: ClusterablePoint[];
    swCount: number;
    soCount: number;
}

/**
 * K-means clustering implementation
 * Groups points into k clusters based on geographic distance
 */
export function kMeansClustering(
    points: ClusterablePoint[],
    k: number,
    maxIterations: number = 100
): Cluster[] {
    if (points.length === 0) return [];
    if (k <= 0) return [];
    if (k >= points.length) {
        // Each point is its own cluster
        return points.map(p => ({
            centroid: p.coordinates,
            points: [p],
            swCount: p.disabilityType === "Sw" ? 1 : 0,
            soCount: p.disabilityType === "So" ? 1 : 0,
        }));
    }

    // Initialize centroids using k-means++ for better initial placement
    const centroids = initializeCentroids(points, k);

    let clusters: Cluster[] = [];
    let prevAssignments: number[] = [];

    for (let iteration = 0; iteration < maxIterations; iteration++) {
        // Assign points to nearest centroid
        const assignments = points.map(point =>
            findNearestCentroid(point.coordinates, centroids)
        );

        // Check for convergence
        if (arraysEqual(assignments, prevAssignments)) {
            break;
        }
        prevAssignments = assignments;

        // Create clusters from assignments
        clusters = centroids.map((centroid, i) => ({
            centroid,
            points: points.filter((_, idx) => assignments[idx] === i),
            swCount: 0,
            soCount: 0,
        }));

        // Update centroids
        clusters.forEach((cluster, i) => {
            if (cluster.points.length > 0) {
                centroids[i] = calculateCentroid(cluster.points);
                cluster.swCount = cluster.points.filter(p => p.disabilityType === "Sw").length;
                cluster.soCount = cluster.points.filter(p => p.disabilityType === "So").length;
            }
        });
    }

    // Final cluster creation
    const finalClusters = centroids.map((centroid, i) => {
        const clusterPoints = points.filter((_, idx) => prevAssignments[idx] === i);
        return {
            centroid,
            points: clusterPoints,
            swCount: clusterPoints.filter(p => p.disabilityType === "Sw").length,
            soCount: clusterPoints.filter(p => p.disabilityType === "So").length,
        };
    });

    // Remove empty clusters
    return finalClusters.filter(c => c.points.length > 0);
}

/**
 * Initialize centroids using k-means++ algorithm
 */
function initializeCentroids(points: ClusterablePoint[], k: number): Coordinates[] {
    const centroids: Coordinates[] = [];

    // Pick first centroid randomly
    const firstIdx = Math.floor(Math.random() * points.length);
    centroids.push({ ...points[firstIdx].coordinates });

    // Pick remaining centroids with probability proportional to distance
    while (centroids.length < k) {
        const distances = points.map(point => {
            const minDist = Math.min(...centroids.map(c =>
                haversineDistance(point.coordinates, c)
            ));
            return minDist * minDist; // Squared distance for probability weighting
        });

        const totalDist = distances.reduce((a, b) => a + b, 0);
        let random = Math.random() * totalDist;

        for (let i = 0; i < points.length; i++) {
            random -= distances[i];
            if (random <= 0) {
                centroids.push({ ...points[i].coordinates });
                break;
            }
        }
    }

    return centroids;
}

/**
 * Find the index of the nearest centroid to a point
 */
function findNearestCentroid(point: Coordinates, centroids: Coordinates[]): number {
    let minDist = Infinity;
    let minIdx = 0;

    centroids.forEach((centroid, i) => {
        const dist = haversineDistance(point, centroid);
        if (dist < minDist) {
            minDist = dist;
            minIdx = i;
        }
    });

    return minIdx;
}

/**
 * Calculate the centroid of a set of points
 */
function calculateCentroid(points: ClusterablePoint[]): Coordinates {
    const sum = points.reduce(
        (acc, p) => ({
            lat: acc.lat + p.coordinates.lat,
            lng: acc.lng + p.coordinates.lng,
        }),
        { lat: 0, lng: 0 }
    );

    return {
        lat: sum.lat / points.length,
        lng: sum.lng / points.length,
    };
}

/**
 * Check if two arrays are equal
 */
function arraysEqual(a: number[], b: number[]): boolean {
    if (a.length !== b.length) return false;
    return a.every((v, i) => v === b[i]);
}

/**
 * Re-cluster if capacity constraints are violated
 * Returns true if valid, false if needs more vehicles
 */
export function validateClusterCapacity(
    cluster: Cluster,
    swCapacity: number,
    soCapacity: number
): boolean {
    return cluster.swCount <= swCapacity && cluster.soCount <= soCapacity;
}

/**
 * Split an over-capacity cluster into two
 */
export function splitCluster(cluster: Cluster): Cluster[] {
    if (cluster.points.length <= 1) {
        return [cluster];
    }

    return kMeansClustering(cluster.points, 2);
}

File: uniride_core/algorithms/clustering.py
"""Core clustering primitives for cluster-first routing pipelines."""

from __future__ import annotations

import random
from dataclasses import dataclass
from typing import List, Tuple

from uniride_core.algorithms.distance import haversine_distance


@dataclass
class Point:
    """Geographic point with student capacity category metadata."""
    id: str
    lat: float
    lng: float
    disability_type: str
    location_code: str = ""


@dataclass
class Cluster:
    """Cluster of student points."""
    centroid: Tuple[float, float]
    points: List[Point]
    sw_count: int
    so_count: int


def calculate_centroid(points: List[Point]) -> Tuple[float, float]:
    if not points:
        return (0.0, 0.0)
    return (
        sum(point.lat for point in points) / len(points),
        sum(point.lng for point in points) / len(points),
    )


def initialize_centroids_kmeans_plus_plus(points: List[Point], k: int) -> List[Tuple[float, float]]:
    if k <= 0 or not points:
        return []

    centroids = []
    first_idx = random.randint(0, len(points) - 1)
    centroids.append((points[first_idx].lat, points[first_idx].lng))

    while len(centroids) < k:
        distances = []
        for point in points:
            min_dist = min(
                haversine_distance(point.lat, point.lng, centroid[0], centroid[1]) ** 2
                for centroid in centroids
            )
            distances.append(min_dist)

        total_dist = sum(distances)
        if total_dist == 0:
            break

        threshold = random.random() * total_dist
        cumulative = 0.0
        for idx, distance in enumerate(distances):
            cumulative += distance
            if cumulative >= threshold:
                centroids.append((points[idx].lat, points[idx].lng))
                break

    return centroids


def kmeans_clustering(points: List[Point], k: int, max_iterations: int = 100) -> List[Cluster]:
    if not points or k <= 0:
        return []

    if k >= len(points):
        return [
            Cluster(
                centroid=(point.lat, point.lng),
                points=[point],
                sw_count=1 if point.disability_type == "Sw" else 0,
                so_count=1 if point.disability_type == "So" else 0,
            )
            for point in points
        ]

    centroids = initialize_centroids_kmeans_plus_plus(points, k)
    prev_assignments = [-1] * len(points)

    for _ in range(max_iterations):
        assignments = []
        for point in points:
            min_idx = 0
            min_dist = float("inf")
            for idx, centroid in enumerate(centroids):
                dist = haversine_distance(point.lat, point.lng, centroid[0], centroid[1])
                if dist < min_dist:
                    min_dist = dist
                    min_idx = idx
            assignments.append(min_idx)

        if assignments == prev_assignments:
            break
        prev_assignments = assignments

        for idx in range(k):
            cluster_points = [points[p_idx] for p_idx in range(len(points)) if assignments[p_idx] == idx]
            if cluster_points:
                centroids[idx] = calculate_centroid(cluster_points)

    clusters = []
    for idx in range(k):
        cluster_points = [points[p_idx] for p_idx in range(len(points)) if prev_assignments[p_idx] == idx]
        if cluster_points:
            clusters.append(
                Cluster(
                    centroid=centroids[idx],
                    points=cluster_points,
                    sw_count=sum(1 for point in cluster_points if point.disability_type == "Sw"),
                    so_count=sum(1 for point in cluster_points if point.disability_type == "So"),
                )
            )
    return clusters


def validate_cluster_capacity(cluster: Cluster, sw_capacity: int, so_capacity: int) -> bool:
    return cluster.sw_count <= sw_capacity and cluster.so_count <= so_capacity


def split_cluster(cluster: Cluster) -> List[Cluster]:
    if len(cluster.points) <= 1:
        return [cluster]
    return kmeans_clustering(cluster.points, 2)


__all__ = [
    "Point",
    "Cluster",
    "calculate_centroid",
    "initialize_centroids_kmeans_plus_plus",
    "kmeans_clustering",
    "validate_cluster_capacity",
    "split_cluster",
]
